File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    right_num, left_num = [], []
    right_set, left_set = set(), set()

    for top in topping:
        right_set.add(top)
        right_num.append(len(right_set))

    for top in topping[::-1]:
        left_set.add(top)
        left_num.append(len(left_set))

    len_left = len(left_num)
    answer = [i for i in range(len(right_num) - 1) if right_num[i] == left_num[len_left - i - 2]]

    return len(answer)

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

public class Solution {
    public int solution(int[] topping) {
        int answer = 0;

        Map<Integer, Integer> rightMap = new HashMap<>();
        Set<Integer> leftSet = new HashSet<>();

        for (int t : topping) {
            rightMap.put(t, rightMap.getOrDefault(t, 0) + 1);
        }

        for (int i = 0; i < topping.length; i++) {
            int t = topping[i];
            
            leftSet.add(t);

            rightMap.put(t, rightMap.get(t) - 1);
            if (rightMap.get(t) == 0) {
                rightMap.remove(t);
            }

            if (leftSet.size() == rightMap.size()) {
                answer++;
            }
        }

        return answer;
    }
}